File: forge/features/auth/templates/platform_auth_sdk_node/node/files/sdks/platform-auth-node/src/trust.ts
/**
 * Per-tenant issuer trust map.
 *
 * Mirrors Python platform_auth.trust: each tenant declares which OIDC
 * issuer should sign tokens for it. Verification rejects tokens whose
 * ``iss`` claim doesn't match the tenant's expected issuer (defends
 * against a compromised second-issuer that could otherwise mint valid
 * tokens for any tenant).
 *
 * The trust map is also where suspended-tenant gating happens:
 * setting ``suspended: true`` blocks verification for that tenant
 * regardless of token validity.
 */

/** Per-tenant trust record. */
export interface TenantTrust {
  /** Issuer URL whose tokens are accepted for this tenant. */
  readonly expectedIssuer: string;
  /** When true, all tokens for this tenant are rejected (TenantSuspended). */
  readonly suspended: boolean;
}

/** Async lookup interface for tenant → trust mapping. */
export interface IssuerTrustMap {
  /** Returns the trust record for ``tenantId``, or ``null`` if unknown. */
  get(tenantId: string): Promise<TenantTrust | null>;
}

/**
 * In-memory map for fixed configurations. Suitable for single-tenant
 * deployments and dev/test fixtures. Multi-tenant production typically
 * fronts this with a Redis or DB-backed implementation.
 */
export class InMemoryIssuerTrustMap implements IssuerTrustMap {
  private readonly tenants: ReadonlyMap<string, TenantTrust>;

  constructor(tenants: Iterable<readonly [string, TenantTrust]> | Record<string, TenantTrust>) {
    const entries = Symbol.iterator in Object(tenants)
      ? (tenants as Iterable<readonly [string, TenantTrust]>)
      : Object.entries(tenants as Record<string, TenantTrust>);
    this.tenants = new Map(entries);
  }

  async get(tenantId: string): Promise<TenantTrust | null> {
    return this.tenants.get(tenantId) ?? null;
  }
}

/**
 * Wraps any IssuerTrustMap with an LRU cache to amortize lookups.
 * The wrapped store sees one call per cache miss / TTL expiry; the
 * AuthGuard hot path reads from memory.
 *
 * Cache invalidation is TTL-only — a tenant whose trust changes
 * upstream (suspension, issuer migration) takes up to ``ttlMs`` to
 * propagate. Operators that need immediate revocation should clear
 * the cache via ``invalidate(tenantId)`` from the same process that
 * mutates the upstream store.
 */
export class CachingIssuerTrustMap implements IssuerTrustMap {
  private readonly inner: IssuerTrustMap;
  private readonly ttlMs: number;
  private readonly cache: Map<string, { value: TenantTrust | null; expiresAt: number }> = new Map();

  constructor(inner: IssuerTrustMap, ttlMs = 60_000) {
    if (ttlMs <= 0) {
      throw new Error("ttlMs must be positive");
    }
    this.inner = inner;
    this.ttlMs = ttlMs;
  }

  async get(tenantId: string): Promise<TenantTrust | null> {
    const now = Date.now();
    const cached = this.cache.get(tenantId);
    if (cached && cached.expiresAt > now) {
      return cached.value;
    }
    const fresh = await this.inner.get(tenantId);
    this.cache.set(tenantId, { value: fresh, expiresAt: now + this.ttlMs });
    return fresh;
  }

  invalidate(tenantId: string): void {
    this.cache.delete(tenantId);
  }
}

File: forge/features/auth/templates/platform_auth_sdk_rust/rust/files/sdks/platform-auth-rs/src/trust.rs
//! Per-tenant issuer trust map.
//!
//! Mirrors Python `platform_auth.trust` and `@forge/platform-auth-node`'s
//! `trust.ts`: each tenant declares which OIDC issuer should sign its
//! tokens. Verification rejects tokens whose `iss` claim doesn't match
//! the tenant's expected issuer (defends against a compromised second
//! issuer that could otherwise mint valid tokens for any tenant).
//!
//! The trust map is also where suspended-tenant gating happens:
//! setting `suspended: true` blocks verification regardless of token
//! validity.

use std::{
    collections::HashMap,
    sync::Arc,
    time::{Duration, Instant},
};

use async_trait::async_trait;
use tokio::sync::RwLock;

/// Per-tenant trust record.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TenantTrust {
    /// Issuer URL whose tokens are accepted for this tenant.
    pub expected_issuer: String,
    /// When true, all tokens for this tenant are rejected (TenantSuspended).
    pub suspended: bool,
}

/// Async lookup interface for tenant → trust mapping.
#[async_trait]
pub trait IssuerTrustMap: Send + Sync {
    async fn get(&self, tenant_id: &str) -> Option<TenantTrust>;
}

/// In-memory map for fixed configurations. Suitable for single-tenant
/// deployments and dev/test fixtures. Multi-tenant production typically
/// fronts this with a Redis or DB-backed implementation.
pub struct InMemoryIssuerTrustMap {
    tenants: HashMap<String, TenantTrust>,
}

impl InMemoryIssuerTrustMap {
    pub fn new() -> Self {
        Self {
            tenants: HashMap::new(),
        }
    }

    pub fn with_tenants<I>(tenants: I) -> Self
    where
        I: IntoIterator<Item = (String, TenantTrust)>,
    {
        Self {
            tenants: tenants.into_iter().collect(),
        }
    }

    pub fn insert(&mut self, tenant_id: String, trust: TenantTrust) {
        self.tenants.insert(tenant_id, trust);
    }
}

impl Default for InMemoryIssuerTrustMap {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl IssuerTrustMap for InMemoryIssuerTrustMap {
    async fn get(&self, tenant_id: &str) -> Option<TenantTrust> {
        self.tenants.get(tenant_id).cloned()
    }
}

/// Wraps any IssuerTrustMap with a TTL cache to amortize lookups.
///
/// Cache invalidation is TTL-only — a tenant whose trust changes
/// upstream (suspension, issuer migration) takes up to `ttl` to
/// propagate. For immediate revocation, call `invalidate(tenant_id)`
/// from the same process that mutates the upstream store.
pub struct CachingIssuerTrustMap {
    inner: Arc<dyn IssuerTrustMap>,
    ttl: Duration,
    cache: RwLock<HashMap<String, CacheEntry>>,
}

#[derive(Clone)]
struct CacheEntry {
    value: Option<TenantTrust>,
    expires_at: Instant,
}

impl CachingIssuerTrustMap {
    pub fn new(inner: Arc<dyn IssuerTrustMap>, ttl: Duration) -> Self {
        assert!(!ttl.is_zero(), "ttl must be positive");
        Self {
            inner,
            ttl,
            cache: RwLock::new(HashMap::new()),
        }
    }

    pub async fn invalidate(&self, tenant_id: &str) {
        self.cache.write().await.remove(tenant_id);
    }
}

#[async_trait]
impl IssuerTrustMap for CachingIssuerTrustMap {
    async fn get(&self, tenant_id: &str) -> Option<TenantTrust> {
        // Fast path: hold the read lock just long enough to check.
        {
            let cache = self.cache.read().await;
            if let Some(entry) = cache.get(tenant_id) {
                if entry.expires_at > Instant::now() {
                    return entry.value.clone();
                }
            }
        }
        // Slow path: refresh under write lock. Double-check so two
        // concurrent misses on the same tenant don't both fetch.
        let mut cache = self.cache.write().await;
        if let Some(entry) = cache.get(tenant_id) {
            if entry.expires_at > Instant::now() {
                return entry.value.clone();
            }
        }
        let fresh = self.inner.get(tenant_id).await;
        cache.insert(
            tenant_id.to_string(),
            CacheEntry {
                value: fresh.clone(),
                expires_at: Instant::now() + self.ttl,
            },
        );
        fresh
    }
}
